File: lever/src/main/java/com/pin/lever/Stats.java
package com.pin.lever;

import android.graphics.Bitmap;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;
import android.os.Message;

import static android.os.Process.THREAD_PRIORITY_BACKGROUND;

class Stats {
    private static final int CACHE_HIT = 0;
    private static final int CACHE_MISS = 1;
    private static final int BITMAP_DECODE_FINISHED = 2;
    private static final int BITMAP_TRANSFORMED_FINISHED = 3;
    private static final int DOWNLOAD_FINISHED = 4;

    private static final String STATS_THREAD_NAME = Utils.THREAD_PREFIX + "Stats";

    final HandlerThread statsThread;
    final Cache cache;
    final Handler handler;

    long cacheHits;
    long cacheMisses;
    long totalDownloadSize;
    long totalOriginalBitmapSize;
    long totalTransformedBitmapSize;
    long averageDownloadSize;
    long averageOriginalBitmapSize;
    long averageTransformedBitmapSize;
    int downloadCount;
    int originalBitmapCount;
    int transformedBitmapCount;

    Stats(Cache cache) {
        this.cache = cache;
        this.statsThread = new HandlerThread(STATS_THREAD_NAME, THREAD_PRIORITY_BACKGROUND);
        this.statsThread.start();
        Utils.flushStackLocalLeaks(statsThread.getLooper());
        this.handler = new StatsHandler(statsThread.getLooper(), this);
    }

    void dispatchBitmapDecoded(Bitmap bitmap) {
        processBitmap(bitmap, BITMAP_DECODE_FINISHED);
    }

    void dispatchBitmapTransformed(Bitmap bitmap) {
        processBitmap(bitmap, BITMAP_TRANSFORMED_FINISHED);
    }

    void dispatchDownloadFinished(long size) {
        handler.sendMessage(handler.obtainMessage(DOWNLOAD_FINISHED, size));
    }

    void dispatchCacheHit() {
        handler.sendEmptyMessage(CACHE_HIT);
    }

    void dispatchCacheMiss() {
        handler.sendEmptyMessage(CACHE_MISS);
    }

    void shutdown() {
        statsThread.quit();
    }

    void performCacheHit() {
        cacheHits++;
    }

    void performCacheMiss() {
        cacheMisses++;
    }

    void performDownloadFinished(Long size) {
        downloadCount++;
        totalDownloadSize += size;
        averageDownloadSize = getAverage(downloadCount, totalDownloadSize);
    }

    void performBitmapDecoded(long size) {
        originalBitmapCount++;
        totalOriginalBitmapSize += size;
        averageOriginalBitmapSize = getAverage(originalBitmapCount, totalOriginalBitmapSize);
    }

    void performBitmapTransformed(long size) {
        transformedBitmapCount++;
        totalTransformedBitmapSize += size;
        averageTransformedBitmapSize = getAverage(originalBitmapCount, totalTransformedBitmapSize);
    }

    StatsSnapshot createSnapshot() {
        return new StatsSnapshot(cache.maxSize(), cache.size(), cacheHits, cacheMisses,
                totalDownloadSize, totalOriginalBitmapSize, totalTransformedBitmapSize, averageDownloadSize,
                averageOriginalBitmapSize, averageTransformedBitmapSize, downloadCount, originalBitmapCount,
                transformedBitmapCount, System.currentTimeMillis());
    }

    private void processBitmap(Bitmap bitmap, int what) {
        // Never send bitmaps to the handler as they could be recycled before we process them.
        int bitmapSize = Utils.getBitmapBytes(bitmap);
        handler.sendMessage(handler.obtainMessage(what, bitmapSize, 0));
    }

    private static long getAverage(int count, long totalSize) {
        return totalSize / count;
    }

    private static class StatsHandler extends Handler {

        private final Stats stats;

        StatsHandler(Looper looper, Stats stats) {
            super(looper);
            this.stats = stats;
        }

        @Override public void handleMessage(final Message msg) {
            switch (msg.what) {
                case CACHE_HIT:
                    stats.performCacheHit();
                    break;
                case CACHE_MISS:
                    stats.performCacheMiss();
                    break;
                case BITMAP_DECODE_FINISHED:
                    stats.performBitmapDecoded(msg.arg1);
                    break;
                case BITMAP_TRANSFORMED_FINISHED:
                    stats.performBitmapTransformed(msg.arg1);
                    break;
                case DOWNLOAD_FINISHED:
                    stats.performDownloadFinished((Long) msg.obj);
                    break;
                default:
                    Lever.HANDLER.post(new Runnable() {
                        @Override public void run() {
                            throw new AssertionError("Unhandled stats message." + msg.what);
                        }
                    });
            }
        }
    }
}

File: lever/src/main/java/com/pin/lever/Stats.kt
package com.pin.lever

import android.graphics.Bitmap
import android.os.*
import com.pin.lever.Lever

class Stats(val cache: Cache) {
    val statsThread: HandlerThread
    val handler: Handler
    var cacheHits: Long = 0
    var cacheMisses: Long = 0
    var totalDownloadSize: Long = 0
    var totalOriginalBitmapSize: Long = 0
    var totalTransformedBitmapSize: Long = 0
    var averageDownloadSize: Long = 0
    var averageOriginalBitmapSize: Long = 0
    var averageTransformedBitmapSize: Long = 0
    var downloadCount = 0
    var originalBitmapCount = 0
    var transformedBitmapCount = 0
    fun dispatchBitmapDecoded(bitmap: Bitmap) {
        processBitmap(bitmap, BITMAP_DECODE_FINISHED)
    }

    fun dispatchBitmapTransformed(bitmap: Bitmap) {
        processBitmap(bitmap, BITMAP_TRANSFORMED_FINISHED)
    }

    fun dispatchDownloadFinished(size: Long) {
        handler.sendMessage(
            handler.obtainMessage(
                DOWNLOAD_FINISHED,
                size
            )
        )
    }

    fun dispatchCacheHit() {
        handler.sendEmptyMessage(CACHE_HIT)
    }

    fun dispatchCacheMiss() {
        handler.sendEmptyMessage(CACHE_MISS)
    }

    fun shutdown() {
        statsThread.quit()
    }

    fun performCacheHit() {
        cacheHits++
    }

    fun performCacheMiss() {
        cacheMisses++
    }

    fun performDownloadFinished(size: Long) {
        downloadCount++
        totalDownloadSize += size
        averageDownloadSize =
            getAverage(downloadCount, totalDownloadSize)
    }

    fun performBitmapDecoded(size: Long) {
        originalBitmapCount++
        totalOriginalBitmapSize += size
        averageOriginalBitmapSize =
            getAverage(originalBitmapCount, totalOriginalBitmapSize)
    }

    fun performBitmapTransformed(size: Long) {
        transformedBitmapCount++
        totalTransformedBitmapSize += size
        averageTransformedBitmapSize = getAverage(
            originalBitmapCount,
            totalTransformedBitmapSize
        )
    }

    fun createSnapshot(): StatsSnapshot {
        return StatsSnapshot(
            cache.maxSize(),
            cache.size(),
            cacheHits,
            cacheMisses,
            totalDownloadSize,
            totalOriginalBitmapSize,
            totalTransformedBitmapSize,
            averageDownloadSize,
            averageOriginalBitmapSize,
            averageTransformedBitmapSize,
            downloadCount,
            originalBitmapCount,
            transformedBitmapCount,
            System.currentTimeMillis()
        )
    }

    private fun processBitmap(
        bitmap: Bitmap,
        what: Int
    ) { // Never send bitmaps to the handler as they could be recycled before we process them.
        val bitmapSize = Utils.getBitmapBytes(bitmap)
        handler.sendMessage(handler.obtainMessage(what, bitmapSize, 0))
    }

    private class StatsHandler internal constructor(looper: Looper?, private val stats: Stats) :
        Handler(looper) {
        override fun handleMessage(msg: Message) {
            when (msg.what) {
                CACHE_HIT -> stats.performCacheHit()
                CACHE_MISS -> stats.performCacheMiss()
                BITMAP_DECODE_FINISHED -> stats.performBitmapDecoded(
                    msg.arg1.toLong()
                )
                BITMAP_TRANSFORMED_FINISHED -> stats.performBitmapTransformed(
                    msg.arg1.toLong()
                )
                DOWNLOAD_FINISHED -> stats.performDownloadFinished(msg.obj as Long)
                else -> Lever.HANDLER.post { throw AssertionError("Unhandled stats message." + msg.what) }
            }
        }

    }

    companion object {
        private const val CACHE_HIT = 0
        private const val CACHE_MISS = 1
        private const val BITMAP_DECODE_FINISHED = 2
        private const val BITMAP_TRANSFORMED_FINISHED = 3
        private const val DOWNLOAD_FINISHED = 4
        private const val STATS_THREAD_NAME = Utils.THREAD_PREFIX + "Stats"
        private fun getAverage(count: Int, totalSize: Long): Long {
            return totalSize / count
        }
    }

    init {
        statsThread = HandlerThread(
            STATS_THREAD_NAME,
            Process.THREAD_PRIORITY_BACKGROUND
        )
        statsThread.start()
        Utils.flushStackLocalLeaks(statsThread.looper)
        handler = StatsHandler(statsThread.looper, this)
    }
}